src/main/java/org/gnieh/logback/config/ConfigConfigurator.java

```diff
@@ -129,67 +111,59 @@ public void configure(LoggerContext loggerContext) {
-		ConfigPropertySetter propertySetter = new ConfigPropertySetter(beanCache, filter);
-		for (String key : config.withoutPath("class").root().keySet()) {
-			propertySetter.setProperty(key, config);
-		}
-
-		filter.setContext(loggerContext);
-		filter.start();
-
-		return filter;
-	}
-
-	private Encoder<?> configureEncoder(LoggerContext loggerContext, Config config)
-			throws ClassNotFoundException, InstantiationException, IllegalAccessException {
-		Class<?> clazz = Class.forName(config.getString("class"));
-		Encoder<?> encoder = (Encoder<?>) clazz.newInstance();
-
-		// for each key appearing in the encoder configuration, check whether
-		// it is a bean property and call the correct method on the encoder
-		ConfigPropertySetter propertySetter = new ConfigPropertySetter(beanCache, encoder);
-		for (String key : config.withoutPath("class").root().keySet()) {
-			if ("layout".equals(key)) {
-				Layout<?> layout = configureLayout(loggerContext, config.getConfig("layout"));
-				propertySetter.setProperty(key, layout);
-			} else {
-				propertySetter.setProperty(key, config);
+	/**
+	 * Configure an object of a given class.
+	 * 
+	 * @param loggerContext
+	 *            the context to assign to this object if it is
+	 *            {@link ContextAwareBase}
+	 * @param clazz
+	 *            the class to instantiate
+	 * @param config
+	 *            a configuration containing the object's properties - each
+	 *            top-level key except for "class" must have a corresponding
+	 *            setter method, or an adder method in the case of lists
+	 * @param children
+	 *            a list which, if not null, will be filled with any child
+	 *            objects assigned as properties
+	 * @return the object instantiated with all properties assigned
+	 * @throws ReflectiveOperationException
+	 *             if any setter/adder method is missing or if the class cannot
+	 *             be instantiated with a no-argument constructor
+	 */
+	private <T> T configureObject(LoggerContext loggerContext, Class<T> clazz, Config config, List<Object> children) throws ReflectiveOperationException {
+		T object = clazz.newInstance();
+
+		if (object instanceof ContextAwareBase)
+			((ContextAwareBase) object).setContext(loggerContext);
+
+		ConfigPropertySetter propertySetter = new ConfigPropertySetter(beanCache, object);
+		for (Entry<String, ConfigValue> entry : config.withoutPath("class").root().entrySet()) {
+			ConfigValue value = entry.getValue();
+			switch (value.valueType()) {
+			case OBJECT:
+				Config subConfig = config.getConfig("\"" + entry.getKey() + "\"");
+				if (subConfig.hasPath("class")) {
+					Class<?> childClass = Class.forName(subConfig.getString("class"));
+					Object child = this.configureObject(loggerContext, childClass, subConfig, null);
+					String propertyName = NameUtils.toLowerCamelCase(entry.getKey());
+					propertySetter.setProperty(propertyName, child);
```

Comment:
Although it makes sense, it's a bit confusing to use the camel-cased property here, but the original key when calling setProperty(String, Config). Perhaps the ConfigPropertySetter should always take a Config as a parameter, and always set using the original key.

Reply:
But here it is a kind of _raw_ `setProperty` without any knowledge about config. It is used to set an object that may be programmatically constructed without any config.

Reply:
Part of what bothers me is they have an identical name, and the raw version simply has a more generic signature. Maybe it would be clearer if the raw version was renamed + javadoc added to indicate what's expected in the "name" parameter in each case. But even if you agree, that can go into a separate commit & PR.

Reply:
I hesitated to name this method `setRawProperty` so I agree with renaming it. The name should conventionally be `propertyName`

Reply:
Hmmmm... while renaming this I was thinking: why not pass the `Config` in the constructor? It seems valid to assume the same `Config` will be used for all setProperty methods?

Reply:
I am not sure this is what we want, because we may use the setter recursively with sub-config.

Reply:
ok, renamed in new commit